Add support for selected fields to blob get and list

mziccard opened the pull request:
This follows from #193. Main changes are:
- Add `AclField` and `BlobField` classes to `Storage`
- Add `BlobGetOption` class to Storage that supports the `.fields` option
- Use `BlobGetOption` in `Storage.get` instead of `BlobSourceOption`
- Add `.fields` option to `BlobListOption`
- Update `Blob.exists` to use selected fields
- Use `BlobGetOption` in `BatchRequest.toGet` instead of `BlobSourceOption`
- Add unit and integration tests for selected fields

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -69,6 +72,209 @@ String entry() {
+      super(selector);
+    }
+
+    public static BlobField acl(AclField... aclFields) {
```
Did you consider just having an "acl" field in this level (basically allowing with or without acl but not getting part of an ACL)? I do wonder how partial ACL will work with patch (with or without id).

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java:
```diff
@@ -159,7 +185,10 @@ public BlobId id() {
-    return storage.get(info.blobId(), convert(info, options)) != null;
+    int length = options.length;
+    Storage.BlobGetOption[] getOptions = Arrays.copyOf(toGetOptions(info, options), length + 1);
+    getOptions[length] = Storage.BlobGetOption.fields();
```
empty string returns nothing or everything (like `null`)?

aozarov replied:
My bad,  I see now  `Storage.BlobGetOption.fields()` will never be empty..

aozarov replied:
Would be nice to add this optimization also to Bucket.exists

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java:
```diff
@@ -71,6 +73,21 @@ private BlobSourceOption(StorageRpc.Option rpcOption) {
       }
     }
 
+    private Storage.BlobGetOption toGetOption(BlobInfo blobInfo) {
```
Maybe we should rename method above to "toSourceOption" to make method name consistent?

Alternative would be to call this method also `convert` with an extra fields argument.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -806,14 +1063,14 @@ private static void checkContentType(BlobInfo blobInfo) throws IllegalArgumentEx
    * @throws StorageException upon failure
    */
-  BlobInfo get(String bucket, String blob, BlobSourceOption... options);
+  BlobInfo get(String bucket, String blob, BlobGetOption... options);
```
Do you think we should also have a `BucketGetOption`?

aozarov commented:
Looks good!

I think, for simplicity, we should reconsider ACL specific fields (the same way we don't expose owner specific fields). I don't even know how patch would work in such case.

Also, if we do so, I wonder if we could just have an enum instead of the `BlobField` and `EntityField` classes.

Providing an option to select fields for buckets (and having a `BucketGetOption`) would be nice too (and if we go with it, I would also suggest to expose only the high-level fields such as acl, lifecycle, cors and not, e.g. cors/maxAgeSeconds).

mziccard commented:
I addressed the comments. I removed `AclField` so that we only deal with first level fields and made `BlobField` an enum.  I also added support for selected fields to bucket get and list.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/spi/DefaultStorageRpc.java:
```diff
@@ -168,6 +169,7 @@ public StorageObject create(StorageObject storageObject, final InputStream conte
           .setPrefix(PREFIX.getString(options))
           .setMaxResults(MAX_RESULTS.getLong(options))
           .setPageToken(PAGE_TOKEN.getString(options))
+          .setFields(FIELDS.getString(options))
```
I assume we need it also on the buckets get and list operations, right?
Add it to the IT test?

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -69,6 +72,104 @@ String entry() {
     }
   }
 
+  public static abstract class EntityField {
```
We don't need that anymore, right?

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -69,6 +72,104 @@ String entry() {
+  }
+
+  enum BucketField {
+    ID("id"),
```
It would be nice to add to IT test a get with explicitly adding all fields (even if we don't check the response), as a typo will be detected there and thrown as an error.

Same for BlobField.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -69,6 +72,104 @@ String entry() {
+    }
+
+    static String selector(BucketField... fields) {
+      HashSet<String> fieldStrings = Sets.newHashSetWithExpectedSize(fields.length + 2);
```
Nit, fields.length + 1

aozarov commented:
Nice, I like it much better.
Few nits and one potential bug regarding bucket get/listing.

mziccard commented:
I fixed the comments, thanks for the careful review.

mziccard commented:
I will need to rebase this to fix conflicts. Please let me know if you are reviewing this, otherwise I'll rebase it first.

aozarov commented:
@mziccard go ahead and re-base and let me know when ready for re-review.

mziccard commented:
@aozarov Rebased and updated according to your last comment.
